Adds user sync flag to the Okta plugin settings

tcsc opened the pull request:
Adds a version field to the OktaPlugin settings and updates the associated protocol files and tests. 

This is in preparation for adding new behaviour to the Okta plugin, and will allow Teleport to determine if a plugin installation was created by the current version of Teleport (and should get the new behaviour), or an old version (which will get no surprising behavioural changes)

tcsc committed:
Up-revs the Okta plugin settings version

Adds a version field to the OktaPlugin settings and updates the associated
protocol files and tests. This is in preparation for adding new behaviour
to the Okta plugin, and will allow Teleport to determine if a plugin
installation was created by the current version of Teleport (and should
get the new behaviour), or an old version (which will get no surprising
behavioural changes)

changing the behaviour of the Okta plugin depending
if the plugin is created from a current or old version of Teleport.

atburke commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -5367,7 +5367,7 @@ message PluginSpecV1 {
     PluginOpenAISettings openai = 3;
     // Settings for the Okta plugin
-    PluginOktaSettings okta = 4;
+    PluginOktaSettingsV2 okta = 4;
```
This is a breaking change. Can we just extend `PluginOktaSettings` instead of making a new type?

tcsc replied:
Yes, should definitely be possible. I wasn't sure what the criteria for up revving the type itself was - but avoiding doing it if possible seems reasonable.

tcsc committed:
revert structure up-rev

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/okta-version-uprev

tcsc committed:
Fix spelling

r0mant commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -5480,6 +5480,9 @@ message PluginOktaSettings {
   string org_url = 1;
+
+  // Version is the version of the Okta settings encoded in this resource.
+  string version = 2;
```
I thought we would just use V2 when creating an Okta Plugin resource but I can see how it could be confusing if we set V2 on PluginV1 resource.

TBH I think having a separate version for settings is also confusing because it becomes hard to understand what the difference is between different versions. How about we just add another explicit property like "bool user_sync_enabled" or something like that instead?

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/okta-version-uprev

tcsc committed:
Revert to simple flag

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/okta-version-uprev

tcsc commented:
Removed `version` and replaced with simple boolean. Ripped out tests for field upgrade as the value for legacy Plugin CRDs is now the boolean zero value (i.e. `false`)

tcsc committed:
Test tidyup

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/okta-version-uprev

tcsc committed:
Merge branch 'master' into tcsc/okta-version-uprev

fspmarshall commented on api/types/plugin_test.go:
```diff
@@ -201,34 +201,63 @@ func TestPluginOpsgenieValidation(t *testing.T) {
 
+func requireBadParameterWith(msg string) require.ErrorAssertionFunc {
+	return func(t require.TestingT, err error, args ...interface{}) {
+		require.True(t, trace.IsBadParameter(err))
```
```suggestion
		require.True(t, trace.IsBadParameter(err), "error: %v", err)
```

tcsc committed:
Update api/types/plugin_test.go

Co-authored-by: Forrest <30576607+fspmarshall@users.noreply.github.com>

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/okta-version-uprev
